### ISSUE TITLE
Long compilation time

### ISSUE THREAD

**Issue by AviManor** [aws/aws-cdi-sdk#46]
Following the instructions in https://github.com/aws/aws-cdi-sdk/blob/mainline/INSTALL_GUIDE_LINUX.md#build-cdi-libraries-and-test-applications 
https://github.com/aws/aws-cdi-sdk/blob/mainline/INSTALL_GUIDE_LINUX.md#build-cdi-libraries-and-test-applications 
cd aws-cdi-sdk/
make DEBUG=y AWS_SDK=<path to AWS SDK C++>
This operation takes a long time. Is it possible to remove unneeded operations and perform the compilation only?
Avi Manor
TAG VS.
avi@tagvs.com

**Comment by mhhen** [comment]
Avi,

Most of the time is spent build third-party components. I took a look at the [Makefile](https://github.com/aws/aws-cdi-sdk/blob/a0d0cbf499e3c96ce3510810f58ae96bb6350480/Makefile#L342) and the AWS SDK is only building two components and not running the unit test. The additional build time should only occur the first time the the CDI-SDK is built. Afterwards, only files that you have changed should compile and link. The third-party components and related unit tests should not be building every time you change a CDI-SDK specific file. Please update this issue with additional details if the build time still seems to be taking a long time for you.

**Comment by AviManor** [comment]
Hi Mike
Compilation always takes more than 80 seconds:

ubuntu@ip-172-31-0-174:/home/AviManor/aws-cdi-sdk$
ubuntu@ip-172-31-0-174:/home/AviManor/aws-cdi-sdk$ SECONDS=0;make V=1 DEBUG=y AWS_SDK=/home/AviManor/aws-sdk-cpp/ > Make.log 2>&1;echo "Total time $SECONDS Sec."
Total time 81 Sec.
ubuntu@ip-172-31-0-174:/home/AviManor/aws-cdi-sdk$

Avi

**Comment by mhhen** [comment]
Avi,

That is strange. I just made a new clone of the CDI-SDK and AWS-SDK. I tested the following command line on both mainline and the developer_preview branch:

`make DEBUG=y AWS_SDK=../aws-sdk-cpp V=1 lib`

Both took about 80 seconds to build the first time. Afterwards, since nothing changed, running the command immediately completes. I then "touched" one of the source files in the cdi folder and ran the command again. It compiled and completed nearly immediately as well. Without using the "lib" target, the HTML documentation is always regenerated, but that only takes a few seconds. You can also use the "test" target to avoid regenerating the documentation. I was using Amazon Linux 2 (not ubuntu), but that should not make a difference.

Can you please provide some more detail on what files are being compiled? Is it the AWS-SDK? libfabric? Maybe the timestamps on some of the source files are in the future, forcing make to always re-build them.

**Comment by AviManor** [comment]
[Make.log](https://github.com/aws/aws-cdi-sdk/files/6817652/Make.log)
Attached is the compilation log. Ignore the errors because I am making changes to the code.
Can I modify the makefile to generate more information?

**Comment by mhhen** [comment]
Avi,

Thanks for providing the log file. It looks like the AWS-SDK is being recompiled and running all its unit tests. Can you confirm that the CDI-SDK Makefile you are using contains this line in the AWS-SDK build rules section: `-DAUTORUN_UNIT_TESTS=FALSE`

[See details here](https://github.com/aws/aws-cdi-sdk/blob/a0d0cbf499e3c96ce3510810f58ae96bb6350480/Makefile#L345)? 

Also, I would suggest that you delete your existing aws-sdk-cpp folder and clone a new one. See steps [here](https://github.com/aws/aws-cdi-sdk/blob/mainline/INSTALL_GUIDE_LINUX.md#download-aws-sdk).

Basically do this:
```
cd <install_dir>
rm -fr aws-sdk-cpp
git clone -b 1.8.46 https://github.com/aws/aws-sdk-cpp.git
```

**Comment by AviManor** [comment]
Hi Mike
The makefile did not include that line, so I added it now. However, I have noticed that my makefile is different from the one in the link. Attached are my makefile and the original one, you can compare them to see the changes I have made. Please note that the problem existed even before I made those changes.
Renewing aws-sdk-cpp did not solve the problem.
I am working on an instance called "temporary instance #01 receiver" in Frankfurt site. Could it be that the time difference is confusing the makefile?
Avi
[Makefiles.zip](https://github.com/aws/aws-cdi-sdk/files/6822878/Makefiles.zip)

**Comment by AviManor** [comment]
Hi Mike
Found this in the makefile:
### The AWS SDK is a required dependency of the CDI SDK by default for the metrics gathering service and for publishing
### metrics to CloudWatch. If the #defines of both CLOUDWATCH_METRICS_ENABLED and METRICS_GATHERING_SERVICE_ENABLED are
### commented out in src/cdi/configuration.h, this requirement can be removed by commenting out the following line.
require_aws_sdk := yes

After commenting out CLOUDWATCH_METRICS_ENABLED, METRICS_GATHERING_SERVICE_ENABLED and require_aws_sdk := yes the compilation takes only one second.
What am I loosing by doing so?
Please note that the root problem was not solved yet.

**Comment by mhhen** [comment]
Avi,

As long as all the source files reside on the same instance, file timestamps based on region should not be causing this problem. Have you tried using a fresh tree and just building the SDK and its related test applications?

We will be providing a new SDK release at the end of this month. Please try it out using a fresh tree (new aws-cdi-sdk, libfabric, and aws-sdk-cpp) and let us know if you are still experiencing this issue. If you would like to run the same test using the currently published version of our developer_preview branch, I would expect the same result (build is not slow after the first build has completed).

**Comment by mhhen** [comment]
Avi,

Please try out [release 2.3](https://github.com/aws/aws-cdi-sdk/releases/tag/v2.3.0) and see if it resolves this issue.

**Comment by AviManor** [comment]
Hi Mike
Issue was not solved. Had to add this line to the Makefile, and undef CLOUDWATCH_METRICS_ENABLED & METRICS_GATHERING_SERVICE_ENABLED, and use NO_MONITORING=y

`### If the #defines of both CLOUDWATCH_METRICS_ENABLED and METRICS_GATHERING_SERVICE_ENABLED are commented out in src/cdi/configuration.h`
`### This requirement can be removed by building with NO_MONITORING=y (not recommended).`
